cmd/soroban-rpc/internal/events/events.go

```diff
@@ -0,0 +1,219 @@
+package events
+
+import (
+	"io"
+	"math"
+	"sort"
+	"sync"
+
+	"github.com/stellar/go/ingest"
+	"github.com/stellar/go/xdr"
+)
+
+// Cursor represents the position of an event within the sequence of all
+// Soroban events. Soroban events are sorted by ledger sequence, transaction
+// index, operation index, and event index.
+type Cursor struct {
+	// Ledger is the sequence of the ledger which emitted the event.
+	Ledger uint32
+	// Tx is the index of the transaction within the ledger which emitted the event.
+	Tx uint32
+	// Op is the index of the operation within the transaction which emitted the event.
+	Op uint32
+	// Event is the index of the event within in the operation which emitted the event.
+	Event uint32
+}
+
+var (
+	// MinCursor is the smallest possible cursor
+	MinCursor = Cursor{}
+	// MaxCursor is the largest possible cursor
+	MaxCursor = Cursor{
+		Ledger: math.MaxUint32,
+		Tx:     math.MaxUint32,
+		Op:     math.MaxUint32,
+		Event:  math.MaxUint32,
+	}
+)
+
+type event struct {
+	contents xdr.ContractEvent
+	cursor   Cursor
+}
+
+func cmp(a, b uint32) int {
+	if a < b {
+		return -1
+	}
+	if a > b {
+		return 1
+	}
+	return 0
+}
+
+// Cmp compares two cursors.
+// 0 is returned if the c is equal to other.
+// 1 is returned if c is greater than other.
+// -1 is returned if c is less than other.
+func (c Cursor) Cmp(other Cursor) int {
+	if c.Ledger == other.Ledger {
+		if c.Tx == other.Tx {
+			return cmp(c.Event, other.Event)
+		}
+		return cmp(c.Tx, other.Tx)
+	}
+	return cmp(c.Ledger, other.Ledger)
+}
+
+// MemoryStore is an in-memory store of soroban events.
+type MemoryStore struct {
+	lock            sync.RWMutex
+	events          []event
+	length          int
+	start           int
+	retentionWindow uint32
+}
+
+// NewMemoryStore creates a new MemoryStore populated by the given ledgers.
+func NewMemoryStore(passPhrase string, ledgers []xdr.LedgerCloseMeta, retentionWindow uint32) (*MemoryStore, error) {
+	var events []event
+	for _, ledger := range ledgers {
+		reader, err := ingest.NewLedgerTransactionReaderFromLedgerCloseMeta(passPhrase, ledger)
+		if err != nil {
+			return nil, err
+		}
+		ledgerEvents, err := readEvents(reader)
+		if err != nil {
+			return nil, err
+		}
+		events = append(events, ledgerEvents...)
+	}
+
+	return &MemoryStore{
+		events:          events,
+		retentionWindow: retentionWindow,
+		length:          len(events),
+	}, nil
+}
+
+// Scan applies f on all the events occurring in the range of [start, end).
+// The events are processed in sorted ascending Cursor order.
+// If f returns false, the scan terminates early (f will not be applied on
+// remaining events in the range).
+func (m *MemoryStore) Scan(start, end Cursor, f func(Cursor, xdr.ContractEvent) bool) {
+	m.lock.RLock()
+	defer m.lock.RUnlock()
+
+	i := sort.Search(m.length, func(i int) bool {
+		index := (m.start + i) % m.length
+		entry := m.events[index]
+		return entry.cursor.Cmp(start) < 0
+	}) + 1
```

Comment:
If we do @tsachiherman's circular buffer above, do we still need the `Search`? If we assume each index in `m.events` is a ledger, couldn't we directly calculate it something like this:
```go
oldest := m.events[m.start].cursor
if oldest > start {
  return fmt.Errorf("cursor out of range") // too old. Some error message here.
}
i := (start - oldest) % m.length
```